test/e2e/run-all.js

```diff
@@ -30,9 +31,44 @@ const getTestPathsForTestDir = async (testDir) => {
   return testPaths;
 };
 
+// Quality Gate Retries
+const RETRIES_FOR_NEW_OR_CHANGED_TESTS = 5;
+let changedOrNewTests;
+
+/**
+ * Runs the quality gate logic to filter and append changed or new tests if present.
+ *
+ * @param {string} fullTestList - List of test paths to be considered.
+ * @returns {string} The updated full test list.
+ */
+async function applyQualityGate(fullTestList) {
+  let qualityGatedList = fullTestList;
+  changedOrNewTests = await filterE2eChangedFiles();
+  console.log('Changed or new tests', changedOrNewTests);
+
+  if (changedOrNewTests.length > 0) {
+    // Filter to include only the paths present in fullTestList
+    const filteredTests = changedOrNewTests.filter((test) =>
+      fullTestList.includes(test),
+    );
+
+    // If there are any filtered tests, append them to fullTestList
+    if (filteredTests.length > 0) {
+      const filteredTestsString = filteredTests.join('\n');
+      for (let i = 0; i < RETRIES_FOR_NEW_OR_CHANGED_TESTS; i++) {
+        qualityGatedList += `\n${filteredTestsString}`;
+      }
+    }
+  }
+
+  return qualityGatedList;
+}
+
 // For running E2Es in parallel in CI
-function runningOnCircleCI(testPaths) {
-  const fullTestList = testPaths.join('\n');
+async function runningOnCircleCI(testPaths) {
+  let fullTestList = testPaths.join('\n');
+  fullTestList = await applyQualityGate(fullTestList);
```

Comment:
Nit: Slight simplification, lets us use `const` which makes tracing this code a bit easier.
```suggestion
  const fullTestList = await applyQualityGate(testPaths.join('\n'));
```

Or alternatively:

```suggestion
  const testPathsList = testPaths.join('\n');
  const fullTestList = await applyQualityGate(testPathsList);
```